File: bindings/csharp/AinosSdk/Models/InferenceChunk.cs
using System.Text.Json.Serialization;

namespace AinosSdk.Models;

/// <summary>
/// A single chunk from a streaming inference response.
/// </summary>
public class InferenceChunk
{
    /// <summary>
    /// The text chunk generated so far.
    /// </summary>
    [JsonPropertyName("chunk")]
    public string Chunk { get; init; } = string.Empty;

    /// <summary>
    /// Whether this is the final chunk in the stream.
    /// </summary>
    [JsonPropertyName("done")]
    public bool Done { get; init; }

    /// <summary>
    /// The index of this chunk in the stream (0-based).
    /// </summary>
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public int Index { get; init; }

    /// <summary>
    /// The model that produced this chunk.
    /// </summary>
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Model { get; init; }

    /// <summary>
    /// Returns a human-readable summary of the chunk.
    /// </summary>
    public override string ToString()
        => $"InferenceChunk {{ Chunk=\"{Chunk[..Math.Min(Chunk.Length, 40)]}..., Done={Done}, Index={Index} }}";
}

File: bindings/java/src/main/java/com/ainos/sdk/models/InferenceChunk.java
package com.ainos.sdk.models;

import java.util.Objects;

/**
 * A single chunk from a streaming inference response.
 * <p>
 * During streaming inference, the daemon sends multiple {@code InferenceChunk}
 * messages, each containing a fragment of the generated text. The final chunk
 * has {@code done} set to {@code true}.
 * <p>
 * This class is immutable and thread-safe.
 */
public final class InferenceChunk {

    private final String chunk;
    private final boolean done;

    /**
     * Constructs a new inference chunk.
     *
     * @param chunk the text fragment for this chunk
     * @param done  whether this is the final chunk
     */
    public InferenceChunk(String chunk, boolean done) {
        this.chunk = Objects.requireNonNull(chunk, "chunk must not be null");
        this.done = done;
    }

    /**
     * Creates a final chunk (done=true) with the given text.
     *
     * @param chunk the final text fragment
     * @return a new InferenceChunk marked as done
     */
    public static InferenceChunk finalChunk(String chunk) {
        return new InferenceChunk(chunk, true);
    }

    /**
     * Creates an intermediate chunk (done=false) with the given text.
     *
     * @param chunk the text fragment
     * @return a new InferenceChunk
     */
    public static InferenceChunk of(String chunk) {
        return new InferenceChunk(chunk, false);
    }

    /** Returns the text fragment for this chunk. */
    public String getChunk() {
        return chunk;
    }

    /**
     * Returns whether this is the final chunk in the stream.
     *
     * @return {@code true} if this is the last chunk
     */
    public boolean isDone() {
        return done;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof InferenceChunk)) return false;
        InferenceChunk that = (InferenceChunk) o;
        return done == that.done && chunk.equals(that.chunk);
    }

    @Override
    public int hashCode() {
        return Objects.hash(chunk, done);
    }

    @Override
    public String toString() {
        return "InferenceChunk{"
                + "chunk='" + chunk + '\''
                + ", done=" + done
                + '}';
    }
}